### PR TITLE
MKL-DNN Data transform doc design

### PR DESCRIPTION
**Pull request by jczaja**
Hi,

I'm one of developers responsible for PaddlePaddle Fluid integration with Intel MKL-DNN.
We would like to start to contribute documents describing various concepts of this integration.

This documentation is relevant to **PaddlePaddle Fluid 1.6 release** . I haven't seen relevant branch so I forked from develop (please advice if this is not proper)

Looking forward to your feedback

### CODE REVIEW

**Comment by jczaja** [comment]
@grygielski Please review

File: doc/fluid/design/mkldnn/data_transformation/data_transform.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+* MKL-DNN kernel is followed by Paddle(CPU) kernel
+* MKL-DNN kernel is followed by fetch operator
+
+
+![](images/data_transform.svg)
+
+
+### Paddle(CPU) kernel is followed by MKL-DNN kernel
+In a situation when Paddle(CPU) kernel finished execution, its outcome is one or many Tensors of Paddle layout. Each of those
+Tensors to be feed into MKL-DNN kernel, needs to be transformed to be of MKL-DNN layout. For this scenario conversion of Paddle Tensor to MKL-DNN Tensor is done by just :
```
Comment (grygielski) [inline doc/fluid/design/mkldnn/data_transformation/data_transform.md]: I think `:` isn't necessary here.

File: doc/fluid/design/mkldnn/data_transformation/data_transform.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+changing layout flag to MKL-DNN and picking MKL-DNN format that match Paddle Tensor rank. This is computationally cheap operation as there is real data rearrangement.
```
Comment (grygielski) [inline doc/fluid/design/mkldnn/data_transformation/data_transform.md]: `as there is real data rearrangement` -> `as there isn't real data rearrangement`

File: doc/fluid/design/mkldnn/data_transformation/data_transform.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+### MKL-DNN kernel is followed by Paddle(CPU) kernel
+In this situation MKL-DNN kernel finished its execution and as a result it produced one or more output Tensors. Each of those Tensors are of MKL-DNN layout and to be fed into Paddle(CPU) kernel,
+they need to be converted into Paddle layout. In a detail MKL-DNN Tensor arrangement (mkl-dnn memory format) is checked if it is compatible with Paddle(CPU) layout and if positive then
+just layout of Tensor is set as Paddle and mkl-dnn format is set to ``undef``. In case when MKL-DNN Tensor data arrangement is not compatible with Paddle layout then actual data arrangement
+is performed. For example MKL-DNN Tensor is 4D and having format ``NCHW16C`` and to convert it into Paddle layout we need to rearrange data to be ``NCHW`` format. To do so
+MKL-DNN Reorder primitive is created that can do data rearrangement. 
+
+This scenario is marked on the picture with regular arrows. Starting from MKL-DNN op on the left side , following regular arrows finishing with fetch op on the right side of picture.
```
Comment (grygielski) [inline doc/fluid/design/mkldnn/data_transformation/data_transform.md]: `finishing with fetch op on the right side of picture.` -> `finishing with Paddle(CPU) op on the right side of picture.`

**Review by grygielski — CHANGES_REQUESTED** [review]
I found some minor mistakes, please check my corrections.

**Review by grygielski — APPROVED** [review]
LGTM

**Comment by jczaja** [comment]
I have rebased branch using newer develop.

@luotao1 could you please review this PR? This is how it looks rendered:
https://github.com/jczaja/FluidDoc/blob/prv-data-transform/doc/fluid/design/mkldnn/data_transformation/data_transform.md

**Comment by luotao1** [comment]
@bingyanghuang Please review it. 
BTW, I could not see the `data_transform.svg`  and `data_transform.dot` in the https://github.com/jczaja/FluidDoc/blob/prv-data-transform/doc/fluid/design/mkldnn/data_transformation/data_transform.md

**Comment by jczaja** [comment]
@luotao1 data_transform.svg is used in data_transform.md . Those files are stored in images and scripts directories (respectively). You can see directory structure here:
https://github.com/jczaja/FluidDoc/tree/prv-data-transform/doc/fluid/design/mkldnn/data_transformation.

**Comment by jczaja** [comment]
@luotao1 Let me know if you have any concerns or further questions about this PR.

**Comment by bingyanghuang** [comment]
@jczaja This document is the first MKL-DNN doc in FLuidDoc which will be published for all the PaddlePaddle users (if it is merged here). If you want to let Paddle users understand MKL-DNN accelerations  in Paddle , I think we should create a more complete and systematic doc including the following explanations: 

- What is MKL-DNN?

- How MKL-DNN integrate to Paddle?(Data transformation can be a part of this section)

- What's kind of performance gain user can get from MKL-DNN(E.g. CNN models acceleration, INT8)

- ......
If you just want to explain what you do for the MKL-DNN data transformation optimization, I think this doc should be merged under paddle/fluid/operators/mkldnn/.

**Comment by luotao1** [comment]
> If you just want to explain what you do for the MKL-DNN data transformation optimization, I think this doc should be merged under paddle/fluid/operators/mkldnn/.

I agree with it.

**Comment by jczaja** [comment]
Ok, I have moved this PR into PaddlePaddle PRs (https://github.com/PaddlePaddle/Paddle/pull/19743)
So, I'm closing this PR

**Comment by luotao1** [comment]
> data_transform.svg is used in data_transform.md . Those files are stored in images and scripts directories (respectively). You can see directory structure here:

Could you directly put a jpg or png in the new doc? We can see the picture in your document.

**Comment by jczaja** [comment]
@luotao1 I'm not sure what you mean. There is a picture (SVG) in MD:
https://github.com/jczaja/Paddle/blob/prv-mkldnn-doc-data-transformation/paddle/fluid/operators/mkldnn/doc/data_transformation/data_transform.md
So when you click that link you should see picture in document. Picture was added with:
``

When you look at data_transform.md then can you see picture rendered?

**Comment by luotao1** [comment]
Got it. Maybe there is something wrong with my chrome, after I change to firefox, I can see the picture.
image

**Comment by grygielski** [comment]
In my Chrome (Version 76.0.3809.132 (Official Build) (64-bit)) everything is fine @luotao1

**Comment by jczaja** [comment]
@luotao1 I could not reopen this PR (I have deleted branch and force replacement of branch is not enough to have PR reopened). So please review #1153

**Pull request closed without merging**